### ISSUE TITLE
completion giving list of all objects

### ISSUE THREAD

**Issue by sevdh** [HenriAugusto/completion-plugin#25]

**Comment by HenriAugusto** [comment]
I can't see to reproduce the issue. Can you give a step-by-step guide?

What system are you on?

**Comment by sevdh** [comment]
I'm on macOS 10.14.6

I have updated the completion plugin via Deken. There could maybe some traces left from the previous install?

When I load Pd I see the following in the Pd window:

[completion-plugin] version 0.47.0-test1
[completion-plugin] finished scanning externals
[completion-plugin] loaded 3194 completions in 210 milliseconds
'completion-plugin.tcl' already loaded, ignoring: '/Users/edwinvanderheide-user/Documents/Pd/externals/completion-plugin/completion-plugin.tcl'

Should the last line indeed be there?

I launch Pd, open a new window with Command+n, I place a new object with Command+1, I type 'sf' in the object and press Command+space. From the list I choose writesf~ by clicking on it and click outside of the object to finish the object creation. Afterwards I place another new object with Command+1, I type 'play' in the object and press Command+space. Now I get a list with, i suppose, all the installed objects including for example links to the completion plugin settings:
<img width="518" alt="Screenshot 2019-08-22 at 06 31 53" src="https://user-images.githubusercontent.com/16959264/63486418-681e9b00-c4e2-11e9-9931-9502a0aadd9f.png">

Sometimes the effect occurs after placing the second object and sometime it only occurs after placing the third object. I haven't experienced it but maybe there are situations where it starts to happen later...

This is the contents of the plugin folder:
<img width="800" alt="Screenshot 2019-08-22 at 06 50 33" src="https://user-images.githubusercontent.com/16959264/63486801-e92a6200-c4e3-11e9-8143-850223fd8e00.png">

These are my settings:
# ===== Completion Plugin =====

# you can change this settings on this file but it is recommended to use the plugin menu while using PD

# hotkey for bringing the completion window.
# for syntax check:
# https://www.tcl.tk/man/tcl8.4/TkCmd/bind.htm
# some examples: "Tab", "Control-Alt-space", "Control"
hotkey Control-space

# maximum lines shown in completion window
max_lines 20

# completion font family
font DejaVu Sans Mono

# fontsize (it usually looks bigger in the completion window so you might need to use a smaller size here)
font_size 12

# The max depth the plugin will scan for externals when descending a folder tree.
max_scan_depth 1

# Auto complete library names. Example:
# On:  "[list-abs/list-compare]"
# Off: "[list-compare]"
auto_complete_libs 1

# ===== Colors =====

# selected completion background color (the color used in the current selected element)
bg #0a85fe

# selected completion background color in the skipping search
skipbg #0ad871

# selected completion background color in the monolithic search
monobg #9832ff

# selected completion foreground color
fg white

# ===== Positions =====

# (vertical) offset for the popup position
# this can be needed if your window manager removes title bar (like awesome, etc)
offset 0

**Comment by HenriAugusto** [comment]
> 'completion-plugin.tcl' already loaded, ignoring: '/Users/edwinvanderheide-user/Documents/Pd/externals/completion-plugin/completion-plugin.tcl'

It do have two plugin copies. How you've installed the plugin the first time?

Maybe you've installed manually the first time in one folder. Then, this time, deken installed it in another folder so you've ended up with two copies.

I've tested having two copies, each in different folder, and i've managed to reproduce this bug.
Also the leftovers `bug1, bug2 bug3` are not included in the 0.47 plugin version so the folder you've posted in the screenshot is an older one.

I'm not sure why it happens when you have two copies because PD should be able to ignore the second plugin but still, this bug doesn't happens with only one copy, at least for me.

I recommend having only a single folder to put all your externals (like `C:\pd-externals`) and using this as your only custom defined path for PD. Then be sure to instruct Deken to install stuff in that folder.

**Comment by HenriAugusto** [comment]
Do a system-wide search for files named `completion-plugin.tcl`.
Keep only the fodler that doesn't contain the _"bug1,bug2, bug3"_ files

 Let me know how it works after removing the duplicate folder.

**Comment by sevdh** [comment]
I have used the Find Any File application to search my drive and there is only one completion-plugin.tcl file on my computer. It’s in /Users/edwinvanderheide-user/Documents/Pd/externals/completion-plugin

I’ve installed the previous version either through Deken or manually (I’m not quite sure but I think it was through Deken).

I’ve updated to the latest version using Deken.

I’m attaching my completion-plugin folder and screenshots of my path and startup settings. Maybe it is occurring because I have both /Users/edwinvanderheide-user/Documents/Pd/externals and /Users/edwinvanderheide-user/Documents/Pd/externals/completion-plugin in my path?

**Comment by sevdh** [comment]
<img width="717" alt="Screenshot 2019-08-22 at 21 34 21" src="https://user-images.githubusercontent.com/16959264/63545285-fe45d600-c561-11e9-896a-4795cdb03c19.png">
<img width="453" alt="Screenshot 2019-08-22 at 21 36 40" src="https://user-images.githubusercontent.com/16959264/63545306-09990180-c562-11e9-9012-77b20305ae2e.png">
[completion-plugin.zip](https://github.com/HenriAugusto/completion-plugin/files/3531707/completion-plugin.zip)

**Comment by HenriAugusto** [comment]
Yes, PD is finding the plugin twice because it finds it after searching those two paths

```
/Users/edwinvanderheide-user/Documents/Pd/externals/
/Users/edwinvanderheide-user/Documents/Pd/externals/completion-plugin
```

That is why i always recommend having only a single search path set.

`/Users/edwinvanderheide-user/Documents/Pd/externals/` in your case. Then just put all your stuff there. And if Deken ask if you wanna add something to the search paths just don't let it.

Anyway if you don't want use the single path for any reason to solve your problem you still need to delete one of the paths.

**Comment by sevdh** [comment]
I see that I’ve been using the search path wrongly!!!
In my case it means that PD would find all the other libraries twice as well… Somehow it is the first time run into this problem and it is an easy mistake to make if you don’t know what to click when Deken asks this question.
I'm happy the problem is solved for me! I think it would be great if the plugin could be made resistant for this because I guess I’m not the only one with overlapping search paths.
Thanks a lot for helping me out!

**Comment by HenriAugusto** [comment]
Yeah, there were always confusion regarding PD searchpaths in our community. It's not something only you did.

Anyway i currently have no idea how this bug is happening.

The relevant plugin-loading code is in [load_plugin_script](https://github.com/pure-data/pure-data/blob/8bc046e972a30eadf1eadea609e9d34426c496ee/tcl/pd-gui.tcl#L747) and [load_startup_plugins](https://github.com/pure-data/pure-data/blob/8bc046e972a30eadf1eadea609e9d34426c496ee/tcl/pd-gui.tcl#L770).

Here's a reduction of what the relevant code on thoses procs looks like (some lines were ommited and are marked with **[...]** )

```TCL
proc load_startup_plugins {} {
    [...]
    # load other installed plugins
    foreach pathdir [concat $::sys_searchpath $::sys_temppath $::sys_staticpath] {
        set dir [file normalize $pathdir]
        if { ! [file isdirectory $dir]} {continue}
        foreach filename [glob -directory $dir -nocomplain -types {f} -- \
                              *-plugin/*-plugin.tcl *-plugin.tcl] {
            set ::current_plugin_loadpath [file dirname $filename]
            load_plugin_script $filename
        }
    }
}
```
```TCL
proc load_plugin_script {filename} {
    global errorInfo

    set basename [file tail $filename]
    if {[lsearch $::loaded_plugins $basename] > -1} {
        ::pdwindow::post [_ "'$basename' already loaded, ignoring: '$filename'\n"]
        return
    }
   [...]
}
```

I'm not sure what the following line does in `load_startup_plugins` 

```TCL
set ::current_plugin_loadpath [file dirname $filename]
```

but it seems to only be used by deken on [pd_deken.tcl](https://github.com/pure-data/pure-data/blob/8bc046e972a30eadf1eadea609e9d34426c496ee/tcl/pd_deken.tcl#L473)

```TCL
if { "" != "$::current_plugin_loadpath" } {
    ::pdwindow::debug [format [_ "\[deken\] deken-plugin.tcl (Pd externals search) loaded from %s." ]  $::current_plugin_loadpath ]
    ::pdwindow::debug "\n"
}
```

So in theory nothing should change...hmm

**Comment by sevdh** [comment]
Another thought: would it be possible that the problem is not the plugin loading twice but that the scan of the completion plugin get’s somehow confused/corrupted because of the overlapping paths?

**Comment by HenriAugusto** [comment]
That is the issue. Looking at the [load_plugin_script](https://github.com/pure-data/pure-data/blob/8bc046e972a30eadf1eadea609e9d34426c496ee/tcl/pd-gui.tcl#L747) and [load_startup_plugins](https://github.com/pure-data/pure-data/blob/8bc046e972a30eadf1eadea609e9d34426c496ee/tcl/pd-gui.tcl#L770) functions i can't find anything wrong.

The [line that loads the plugin](https://github.com/pure-data/pure-data/blob/8bc046e972a30eadf1eadea609e9d34426c496ee/tcl/pd-gui.tcl#L760) is not called when the code detects the plugin is already loaded
```TCL
[catch {uplevel #0 $tclcode} errorname]
```

Of course I'm missing something but, currently, i don't know where to look.

**Comment by HenriAugusto** [comment]
Through issue #30 'ive realized that the plugin is actually using the `::current_plugin_loadpath` variable!

Interestingly:

1. the plugin reads it on initialization, which souldn't be a problem.
   - it reads the .cfg file
   - uses this path to find the objects located in `custom_completions` and `monolithic_objects` folders
2. the plugin reads it at run-time which might be a problem
   - but it only does when opening the help file and writing to the config file.

So while it opens the chance for chaos by reading this variable i don't see how it can affect the search routines, even if somehow we wrote the config file to the wrong place.

Moreover i've tried reproducing this issue again with the plugin v0.47.0 and couldn't. Maybe it was somehow related to #26 ?

**Comment by sevdh** [comment]
v0.47.0 gives still the same issue for me. I’ll do more testing.
